File: day17.lua
#!/usr/bin/env lua

local adv <const> = 0
local bxl <const> = 1
local bst <const> = 2
local jnz <const> = 3
local bxc <const> = 4
local out <const> = 5
local bdv <const> = 6
local cdv <const> = 7

local A <const> = 1
local B <const> = 2
local C <const> = 3

local reg = {}
local prg = {}

local function combo(o)
    if o < 4 then return o end
    if o < 7 then return reg[o - 3] end
    assert(false)
end

local function run(limit)
    local output = {}
    local ip = 0
    local ipn
    limit = limit or math.maxinteger
    while ip >= 0 and ip < #prg and #output < limit do
        assert(ip % 2 == 0)
        local opc, o = prg[ip + 1], prg[ip + 2]
        ipn = ip + 2
        if     opc == adv then reg[A] = reg[A] >> combo(o)
        elseif opc == bxl then reg[B] = reg[B] ~ o
        elseif opc == bst then reg[B] = combo(o) % 8
        elseif opc == jnz then if reg[A] ~= 0 then ipn = o end
        elseif opc == bxc then reg[B] = reg[B] ~ reg[C]
        elseif opc == out then output[#output + 1] = combo(o) % 8
        elseif opc == bdv then reg[B] = reg[A] >> combo(o)
        elseif opc == cdv then reg[C] = reg[A] >> combo(o)
        end
        ip = ipn
    end
    return output
end

local function solve(accum, pos)
    if pos < 1 then return accum end
    for low = 0, 7 do
        local a = (accum << 3) | low
        reg[A] = a
        local x = run(1)[1]
        if x == prg[pos] then
            local result = solve(a, pos - 1)
            if result then return result end
        end
    end
end

local file = io.open("day17.txt")

for line in file:lines() do
    local rs, xs = line:match("Register (%u): (%d+)")
    if not rs then break end
    local r, x = rs:byte(), tonumber(xs)
    --print(string.format("Register %c: %d", r, x))
    reg[r - ("A"):byte() + 1] = x
end

for xs in file:read():gmatch("%d") do
    local x = tonumber(xs)
    --print(string.format("%d,", x))
    prg[#prg + 1] = x
end

local output = run()
print(table.concat(output, ","))

print(solve(0, #prg))

File: day17.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>

#define COUNT(a) (sizeof(a) / sizeof((a)[0]))

enum {
    adv,
    bxl,
    bst,
    jnz,
    bxc,
    out,
    bdv,
    cdv,
};

#define A reg[0]
#define B reg[1]
#define C reg[2]

typedef long long reg_t;

reg_t reg[3];
int prg[16];
int n;

reg_t combo(int o)
{
    if (o < 4) return o;
    if (o < 7) return reg[o - 4];
    assert(false); return 0;
}

int run(int* output, int limit)
{
    int noutput = 0;
    for (int ip = 0, ipn; ip >= 0 && ip < n && noutput < limit; ip = ipn) {
        assert(ip % 2 == 0);
        int opc = prg[ip], o = prg[ip + 1];
        ipn = ip + 2;
        switch (opc) {
        case adv: A = A >> combo(o); break;
        case bxl: B = B ^ o; break;
        case bst: B = combo(o) % 8; break;
        case jnz: if (A) { ipn = o; } break;
        case bxc: B = B ^ C; break;
        case out: output[noutput++] = combo(o) % 8; break;
        case bdv: B = A >> combo(o); break;
        case cdv: C = A >> combo(o); break;
        }
    }
    return noutput;
}

reg_t solve(reg_t accum, int pos)
{
    if (pos < 0) return accum;
    for (int low = 0; low < 8; low++) {
        reg_t a = (accum << 3) | low;
        A = a;
        int x;
        if (run(&x, 1) && x == prg[pos]) {
            reg_t result = solve(a, pos - 1);
            if (result >= 0) return result;
        }
    }
    return -1;
}

int main()
{
    FILE* file = fopen("day17.txt", "r");

    char r;
    for (int x; 2 == fscanf(file, "Register %c: %d\n", &r, &x); ) {
        //printf("Register %c: %d\n", r, x);
        reg[r - 'A'] = x;
    }

    fscanf(file, "Program: ");

    for (int x, y; 2 == fscanf(file, "%d,%d,", &x, &y); ) {
        //printf("%d,%d,\n", x, y);
        prg[n++] = x;
        prg[n++] = y;
    }

    int output[16];
    int noutput = run(output, COUNT(output));
    assert(noutput < COUNT(output));
    for (int i = 0; i < noutput; i++) {
        if (i > 0) printf(",");
        printf("%d", output[i]);
    }
    printf("\n");

    printf("%lld\n", solve(0, n - 1));
}
